### ISSUE TITLE
pi07 low-level planner forward is not byte-equivalent to pi05 when all optional inputs are dropped

### ISSUE THREAD

**Issue by shuheng-liu** [TensorAuto/OpenTau#226]
## Context

When `subgoal_drop_prob = response_drop_prob = metadata_drop_all_prob = metadata_drop_each_prob = 1.0` (e.g. the smoke config below), `PI07LowLevelPlannerFlowMatching.embed_prefix` should produce a prefix that is functionally identical to `PI05FlowMatching.embed_prefix` (continuous state, `predict_response=False`) — same weights → same KV cache → same losses. It does not. Three small divergences in the prefix layout / attention pattern make pi07 emit a different prefix even after #205 landed (which gated the *content* of the optional middle blocks but not the separators around them).

## Repro

`fix/pi07_paligemma_fixes` (currently `9528cfb`). Smoke test: instantiate both policies (`pretrained_path=None`, identical seed), copy `paligemma_with_expert.state_dict()` + the projection layers from pi05 to pi07, cast both to bf16, feed one fixed batch with the same noise / time:

```
camera0/camera1: (B=2, 3, 224, 224)  random in [0, 1]
state:           (B=2, 32)            random in [-1, 1]
actions:         (B=2, 10, 32)        random in [-1, 1]
prompt:          ["pick up the red block"] * 2
no response/subgoal/metadata keys -> all gating False
```

Result on `fix/pi07_paligemma_fixes` head:

```
pi05  MSE=81.16315460  CE=11.03499031   prefix=(2, 681, 2048)
pi07  MSE=81.89469910  CE=11.47132397   prefix=(2, 684, 2048)
diff  MSE=7.31e-01     CE=4.36e-01      +3 prefix tokens in pi07
```

Per-segment byte check (max abs diff between pi05's and pi07's prefix embeddings):

| segment              | pi05 slice  | pi07 slice  | max abs diff |
| -------------------- | ----------- | ----------- | ------------ |
| video / image (T=1)  | `[0:512]`   | `[0:512]`   | **0.000** |
| lang                 | `[512:612]` | `[512:612]` | **0.000** |
| `"State: "`          | `[612:615]` | `[612:615]` | **0.000** |
| state_emb            | `[615:616]` | `[615:616]` | **0.000** |
| pi05 `":\n"` (2)     | `[616:618]` | —           | (different token from pi07's `", "`) |
| pi07 `", "` (2)      | —           | `[616:618]` | |
| **pi07 `";\n "` (3)** | —          | **`[618:621]`** | **— spurious 3 dangling tokens** |
| `"Action: "`         | `[618:621]` | `[621:624]` | **0.000** (after offset) |
| discrete actions (60)| `[621:681]` | `[624:684]` | **0.000** |

The image / video encoder is byte-identical at T=1 (max abs diff = 0 between `embed_image` via the patched `get_image_features` and `SpaceTimeSiglipVideoEncoder.forward`). All shared params line up after the explicit copy. The whole divergence is in `embed_prefix`.

## Three divergences

`src/opentau/policies/pi07_paligemma/low_level_planner/modeling_pi07_low_level.py`, in `PI07LowLevelPlannerFlowMatching.embed_prefix`:

1. **`", "` state-end is unconditionally appended** (currently around line 1212). When all optionals are dropped the comma is a dangling "state, [nothing]" separator that adds 2 real tokens absent from the pi05 layout. Pi05 uses `":\n"` here.

2. **`";\n "` prefix-end is unconditionally appended** (currently around line 1321). When the optional middle list is empty there is no list to terminate; this adds 3 real tokens absent from the pi05 layout.

3. **`"Action: "` indicator att_masks differ from pi05** (currently around line 1390):
   - pi07: `att_masks += [1] + [0] * (N - 1)` — all N indicator tokens form a single bidirectional block.
   - pi05: `att_masks += [1] * N` — each indicator token is its own causal block.

   The embeddings end up byte-identical, but `make_att_2d_masks` produces different attention patterns: in pi07 the cumsum stays at 1 across the whole indicator and every subsequent discrete-action token's cumsum is shifted down by 2. The action expert path doesn't touch the indicator (so MSE matches once 1 + 2 are fixed), but `prefix_out` at the indicator → first-discrete boundary changes, which is where the discrete-action CE loss reads, so CE diverges.

The recent fix #205 correctly gated the *content* blocks (response / subgoal / metadata) to skip when all samples have pad-only masks, but left the surrounding separators ungated. This issue follows up.

## Suggested fix

Compute once at the top of `embed_prefix`:

```python
has_response = (response_tokens is not None and response_masks is not None and response_masks.any())
has_subgoal  = bool(subgoal_images) and bool(subgoal_img_masks) and any(m.any() for m in subgoal_img_masks)
has_metadata = (metadata_tokens is not None and metadata_masks is not None and metadata_masks.any())
has_any_optional = bool(has_response or has_subgoal or has_metadata)
```

Then:

- State-end: `state_end_str = ", " if has_any_optional else ":\n"` and proceed as today.
- Wrap the entire `";\n "` prefix-end block in `if has_any_optional:`.
- Discrete-action indicator: `att_masks += [1] * num_discrete_action_start_embs` (matches pi05).

After these three changes, with the same smoke test:

```
pi05  MSE=81.16315460  CE=11.03499031
pi07  MSE=81.16315460  CE=11.03499031
diff  MSE=0.000000e+00  CE=0.000000e+00

FULL prefix max diff: 0.000e+00, mean: 0.000e+00
att_masks differ at 0 positions, pad_masks differ at 0
```

Byte-identical losses, prefix embeddings, and attention/pad masks.

## Test impact

`tests/policies/test_pi07_paligemma_low_level_planner.py::test_complete_pi07_low_level_pipeline` populates **all** optional blocks, so the `has_any_optional == True` branch is exercised and the prefix length is unchanged from today. The att_masks change for the `"Action: "` indicator does change the cumsum at the indicator → discrete boundary, so any test asserting the exact attention pattern there will need updating (the pad_masks shape is unchanged).

## Out of scope

- The pi07 high-level planner (only the low-level planner was inspected).
- Whether pi05's hardcoded `discrete_action_indicator_max_length = 3` matches the actual paligemma tokenization (it does on `google/paligemma-3b-pt-224`, but worth converting to a tokenizer-derived length for robustness).
- The semantic question of whether pi07 *should* be byte-equivalent to pi05 in the all-dropped case. Filing this assuming yes — it's the cleanest way to reuse pi05_base checkpoints and to keep the two architectures comparable in ablations.

## References

- #205 — gated optional content blocks; this issue is the follow-up for the surrounding separators.
- #182 — pi05 SDPA / fp32 Adam (referenced for the bf16 forward-path conventions used in the repro).
- PR #167 — introduced pi07 paligemma planners.

**Comment by shuheng-liu** [comment]
Should have been fixed by #260
